core/src/main/scala/caliban/parsing/VariablesUpdater.scala

```diff
@@ -50,49 +88,63 @@ object VariablesUpdater {
+                .map(field => coerceValues(v, field.`type`(), rootType).map(k -> _))
+                .getOrElse(IO.succeed(k -> value))
+            }.map(InputValue.ObjectValue(_))
+          case NullValue                      => IO.succeed(NullValue)
+          case v                              =>
+            IO.fail(ValidationError(s"Cannot coerce $v to INPUT_OBJECT", ""))
         }
 
       case __TypeKind.LIST =>
         value match {
           case ListValue(values) =>
             typ.ofType
-              .map(innerType => ListValue(values.map(value => coerceValues(value, innerType, rootType))))
-              .getOrElse(value)
-          case _                 => value
+              .map(innerType => IO.foreach(values)(coerceValues(_, innerType, rootType)).map(ListValue(_)))
+              .getOrElse(IO.succeed(value))
+          case NullValue         => IO.succeed(NullValue)
+          case v                 => IO.fail(ValidationError(s"Cannot coerce $v into LIST", ""))
         }
 
       case __TypeKind.NON_NULL =>
-        typ.ofType
-          .map(innerType => coerceValues(value, innerType, rootType))
-          .getOrElse(value)
+        value match {
+          case NullValue =>
+            IO.fail(
+              ValidationError(
+                s"$value is null",
+                "Arguments can be required. An argument is required if the argument type is non‐null and does not have a default value. Otherwise, the argument is optional."
+              )
+            )
+          case _         =>
+            typ.ofType
+              .map(innerType => coerceValues(value, innerType, rootType))
+              .getOrElse(IO.succeed(value))
+        }
 
       case __TypeKind.ENUM                                 =>
         value match {
-          case StringValue(value) => Value.EnumValue(value)
-          case _                  => value
+          case StringValue(value) => IO.succeed(Value.EnumValue(value))
+          case NullValue          => IO.succeed(NullValue)
+          case v                  => IO.fail(ValidationError(s"Cannot coerce $v into ENUM.", ""))
         }
       case __TypeKind.SCALAR if typ.name.contains("Float") =>
         value match {
-          case IntValue.IntNumber(value)    => Value.FloatValue(value.toDouble)
-          case IntValue.LongNumber(value)   => Value.FloatValue(value.toDouble)
-          case IntValue.BigIntNumber(value) => Value.FloatValue(BigDecimal(value))
-          case _                            => value
+          case NullValue                    => IO.succeed(NullValue)
+          case IntValue.IntNumber(value)    => IO.succeed(Value.FloatValue(value.toDouble))
+          case IntValue.LongNumber(value)   => IO.succeed(Value.FloatValue(value.toDouble))
+          case IntValue.BigIntNumber(value) => IO.succeed(Value.FloatValue(BigDecimal(value)))
+          case v                            => IO.fail(ValidationError(s"Cannot coerce $v into number", ""))
```

Comment:
How should we format coercion errors? With some errors context similar to `Validator.scala`?

Reply:
If the path was available that would be the next UX

Reply:
✅